autogen/agentchat/contrib/compressible_agent.py

```diff
@@ -0,0 +1,392 @@
+            if max_token_allowed - token_used <= 0:
+                # Teminate if no token left.
+                print(
+                    colored(
+                        f'Warning: Terminate Agent "{self.name}" due to no token left for oai reply. max token for {model}: {max_token_allowed}, existed token count: {token_used}',
+                        "yellow",
+                    ),
+                    flush=True,
+                )
+                return True, None
+            return False, None
+
+        # if token_used is less than trigger_count, no compression will be used.
+        if token_used < self.compress_config["trigger_count"]:
+            return False, None
+
+        # 2. mode = "COMPRESS" or mode = "CUSTOMIZED", compress the messages
+        copied_messages = copy.deepcopy(messages)
+        if self.compress_config["mode"] == "COMPRESS":
+            return self.compress_messages(copied_messages)
+        elif self.compress_config["mode"] == "CUSTOMIZED":
+            return self.compress_config["compress_function"](copied_messages)
+        else:
+            raise ValueError(f"Unknown compression mode: {self.compress_config['mode']}")
+
+    def on_oai_token_limit(
+        self,
+        messages: Optional[List[Dict]] = None,
+        sender: Optional[Agent] = None,
+        config: Optional[Any] = None,
+    ) -> Tuple[bool, Union[str, Dict, None]]:
+        """(Experimental) Compress previous messages when a threshold of tokens is reached.
+
+        TODO: async compress
+        TODO: maintain a list for old oai messages (messages before compression)
+        """
+        llm_config = self.llm_config if config is None else config
+        if self.compress_config is False:
+            return False, None
+        if messages is None:
+            messages = self._oai_messages[sender]
+
+        model = llm_config["model"]
+        token_used = self._compute_init_token_count() + count_token(messages, model)
+        final, compressed_messages = self._manage_history_on_token_limit(
+            messages, token_used, get_max_token_limit(model), model
+        )
+
+        # update message history with compressed messages
+        if compressed_messages is not None:
+            to_print = (
+                "Token Count (of msgs after first prompt): Before compression: {} After: {} | "
+                "Total prompt token count after compression: {}".format(
+                    count_token(self._oai_messages[sender][1:], model),
+                    count_token(compressed_messages[1:], model),
+                    count_token(compressed_messages, model) + self._compute_init_token_count(),
+                )
+            )
+            print(colored(to_print, "magenta"), flush=True)
+            print("-" * 80, flush=True)
+
+            self._oai_messages[sender] = compressed_messages
+            if self.compress_config["broadcast"]:
+                sender._oai_messages[self] = copy.deepcopy(compressed_messages)
```

Comment:
This is not right. The message list in self and in sender should not be identical. The role for "user" and "assistant" can be different.
Also, this doesn't work for group chat. Is it documented?

Reply:
Revised.
Yes, it is stated in the notebook.